Fix files.sync exclude dir multiple

gchazot opened the pull request:
This contains some fixing and clean-up of the `files.sync` operation, with most importantly:
* c43c1df + 278d1a8 -> Fixing handling of list in `exclude_dir` argument
* c41490d and 2b90e80 are 2 distinct solutions to handle excluding "sub-directories"

Discussion is needed around the solutions for excluding "sub-directories". To be more explicit, the question is:

What behaviour do we want when the deploy specifies `exclude_dir="__pycache__"` (for example)? Do we want this to be enough to exclude any directory with this name in the `src` tree, or only the one possibly present right under `src` directory? 

Note that a very similar question should apply to the `exclude=...` argument, in order to keep both arguments behaviour consistent.

Commit:
Rename member variables to explicit internal use

Commit:
Extract method to parse local files and directories in tests

Commit:
Add test exposing sync bug with multiple exclude dirs

Commit:
Fix filtering of subdirectories for sync

Removing entries in a list while iterating it is Bad(tm)

Commit:
Expose solution to exclude subdirectories

Commit:
Add explicit test for creation of dest with files.sync

Commit:
Extract variable to make list content more explicit

Commit:
Use os.path to normalise local path manipulations

gchazot commented:
Starting the discussion about the `exclude_dir` behaviour, I think that the current behaviour is already the better one, i.e. the one presented with c41490d80769cb62506c6728321af8f9a3113684. It is a little annoying because it is repeating the argument (`__pycache__` + `*/__pycache__`), but this solution allows to say explicitely with the wildcard that the folder name should be excluded in all subfolders.

The repetition is inherent to the use of `fnmatch` as the pattern handling function and because it doesn't treat `/` as a special character, meaning `*/` does not match "no folder" (i.e. `src`)


The same argument goes against the other solution (2b90e80ed5c3dbf76a4e6fc582d38da4a72a93dd), since there, specifying `__pycache__` is the only way to specify that `src/__pycache__` must be excluded, but it also means that *any* other `__pycache__` folder nested below `src` will also be implicitly excluded

Commit:
Implement new normalised structure to provide mock filesystem data

Commit:
Fix files.sync test on Windows

Fizzadar commented:
I'm a strong agree on the explicit first method. Annoying to specify twice but means you know exactly what the operation is going to do. This also matches the current logic thus avoiding a breaking change.

I think putting an example of the double exclude in the docstring will suffice to ensure this is well documented.

gchazot committed:
Remove leftover debug logging

Commit:
Migrate all tests to new local filesystem mock

Commit:
Remove old filesystem mocking format

Commit:
Migrate Windows test to new filesystem mock format

Commit:
Normalise base path for mocked local files.

This is a superficial patch to make sure internal paths to mocked files and directories are using the current OS format.
However, externally, especially in operation parameters, the paths remain in the user-provided format.
This is useful in tests because it means the same JSON can be used on Unix and Windows to describe tests.

Commit:
Add note in files.sync documentation explaining exclusions

gchazot commented:
I've rebased this on top of `gchazot:rework_filesystem_mock` (#675) to benefit from the new filesystem mock format.

Removed solution 2 commits

Added note to the documentation to explain the subtlety of `exclude_dir`:
![image](https://user-images.githubusercontent.com/8948373/136918690-4835466c-57fe-45c2-b282-847f29774cd6.png)

gchazot commented:
This is ready I think, just depends on #675 being merged first

Fizzadar reviewed (approved):
Thank you @gchazot!
